File: src-tauri/src/dto/novel.rs
use crate::dto::common::ReadPosition;
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "snake_case")]
pub(crate) struct NovelSummary {
    pub source: String,
    pub id: String,
    pub title: String,
    pub cover_url: Option<String>,
    pub author: Option<String>,
    pub status: Option<String>,
    pub updated_at: Option<String>,
    pub description: Option<String>,
    pub tags: Vec<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct NovelChapterSummary {
    pub id: String,
    pub title: String,
    pub sequence: i64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct NovelDetail {
    #[serde(flatten)]
    pub summary: NovelSummary,
    pub chapters: Vec<NovelChapterSummary>,
    pub read_position: Option<ReadPosition>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct NovelChapterContent {
    pub chapter_id: String,
    pub server_chapter_id: String,
    pub title: String,
    pub html: String,
    pub font_url: Option<String>,
    pub read_position: Option<ReadPosition>,
}

File: src/domain/novel.ts
/**
 * Novel-domain DTOs rendered by the frontend.
 *
 * Novel and comic remain separate domains, but their summaries, catalogue
 * entries, reading positions, and reader payloads live here instead of in
 * transport adapters.
 */
import type { ReadPosition } from "./readPosition";

export interface NovelSummary {
  source: string;
  id: string;
  title: string;
  cover_url: string | null;
  author: string | null;
  status: string | null;
  updated_at: string | null;
  description: string | null;
  tags: string[];
}

export interface NovelChapterSummary {
  id: string;
  title: string;
  sequence: number;
}

export interface NovelDetail extends NovelSummary {
  chapters: NovelChapterSummary[];
  readPosition: ReadPosition | null;
}

/** Sanitized novel chapter payload consumed by the text reader. */
export interface NovelChapterContent {
  chapterId: string;
  serverChapterId: string;
  title: string;
  html: string;
  fontUrl: string | null;
  readPosition: ReadPosition | null;
}

export type { DiscoveryList, RecommendBlock } from "./discovery";
